File: android/app/src/main/kotlin/ai/gowtam/oak/ui/MarkdownBlockView.kt
package ai.gowtam.oak.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Constraints
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Renders block-level Markdown — headings, lists, GFM tables, fenced code,
 * blockquotes, thematic breaks — natively, using [MarkdownBlocks.parse] to split the
 * source and [parseInline] to render the bold/italic/`code`/link markup *inside* each
 * block. The block-aware counterpart to [MarkdownText] (which stays inline-only).
 *
 * **Streaming-safe by construction:** the splitter degrades a half-formed trailing
 * block (an open code fence, a header without its separator) to a plain inline-parsed
 * paragraph, so re-rendering on every delta never drops content and never flickers
 * structure it can't yet resolve.
 *
 * Ports `ios/OakApp/UI/MarkdownBlockView.swift`.
 */
@Composable
fun MarkdownBlockView(
    markdown: String,
    modifier: Modifier = Modifier,
) {
    val blocks = MarkdownBlocks.parse(markdown)
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        for (block in blocks) {
            BlockView(block)
        }
    }
}

/**
 * Inline-only Markdown prose — the counterpart to iOS `MarkdownText`. Renders
 * bold/italic/`code`/links into a single wrapping [Text]; block structure is ignored.
 * Use where a field is known to be a single run of prose.
 */
@Composable
fun MarkdownText(
    markdown: String,
    modifier: Modifier = Modifier,
    color: Color = Color.Unspecified,
) {
    val oak = LocalOakColors.current
    Text(
        text = parseInline(markdown, linkColor = oak.azure),
        modifier = modifier,
        color = color,
    )
}

@Composable
private fun BlockView(block: MdBlock) {
    val oak = LocalOakColors.current
    when (block) {
        is MdBlock.Heading -> Text(
            text = parseInline(block.text, linkColor = oak.azure),
            style = headingStyle(block.level),
            color = oak.textStrong,
            modifier = Modifier.fillMaxWidth(),
        )

        is MdBlock.Paragraph -> Text(
            text = parseInline(block.text, linkColor = oak.azure),
            style = MaterialTheme.typography.bodyLarge,
            modifier = Modifier.fillMaxWidth(),
        )

        is MdBlock.MdList -> MarkdownListView(block.items)

        is MdBlock.Table -> MarkdownTableView(block.table)

        is MdBlock.CodeBlock -> MarkdownCodeBlock(block.code)

        is MdBlock.Blockquote -> MarkdownBlockquote(block.text)

        MdBlock.ThematicBreak -> HorizontalDivider(color = oak.border)
    }
}

@Composable
private fun headingStyle(level: Int) = when (level) {
    1 -> MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.SemiBold)
    2 -> MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold)
    3 -> MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold)
    else -> MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold)
}

// ---- Lists ----

@Composable
private fun MarkdownListView(items: List<MdListItem>) {
    val oak = LocalOakColors.current
    Column(verticalArrangement = Arrangement.spacedBy(OakSpacing.xs)) {
        for (item in items) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = (item.level * 18).dp),
                horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm),
            ) {
                Text(
                    text = marker(item),
                    color = oak.textMuted,
                    style = MaterialTheme.typography.bodyLarge,
                    fontFamily = if (item.ordered) FontFamily.Monospace else FontFamily.Default,
                )
                Text(
                    text = parseInline(item.text, linkColor = oak.azure),
                    style = MaterialTheme.typography.bodyLarge,
                    modifier = Modifier.weight(1f),
                )
            }
        }
    }
}

private fun marker(item: MdListItem): String = when {
    item.ordered -> "${item.number}."
    item.level == 0 -> "•" // •
    else -> "◦" // ◦
}

// ---- Code block ----

@Composable
private fun MarkdownCodeBlock(code: String) {
    val oak = LocalOakColors.current
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(OakRadius.md))
            .background(oak.surfaceSunken)
            .border(1.dp, oak.border, RoundedCornerShape(OakRadius.md))
            .horizontalScroll(rememberScrollState())
            .padding(OakSpacing.md),
    ) {
        Text(
            text = code,
            fontFamily = FontFamily.Monospace,
            fontSize = 13.sp,
            color = oak.text,
        )
    }
}

// ---- Blockquote ----

@Composable
private fun MarkdownBlockquote(text: String) {
    val oak = LocalOakColors.current
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(OakSpacing.sm),
    ) {
        Box(
            modifier = Modifier
                .width(3.dp)
                .background(oak.accent.copy(alpha = 0.5f), RoundedCornerShape(2.dp)),
        )
        Text(
            text = parseInline(text, linkColor = oak.azure),
            style = MaterialTheme.typography.bodyLarge,
            color = oak.textMuted,
            modifier = Modifier.weight(1f),
        )
    }
}

// ---- Table ----

/**
 * A GFM table rendered as a native grid, horizontally scrollable so a wide table
 * never forces the page to scroll sideways. Column widths size to content (capped)
 * and every cell in a visual column shares that width, so rows line up. Header row
 * gets a wash + underline; body rows zebra-stripe.
 */
@Composable
private fun MarkdownTableView(table: MdTable) {
    val oak = LocalOakColors.current
    val cols = maxOf(table.header.size, 1)
    val allRows = remember(table) { listOf(table.header) + table.rows }

    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(OakRadius.md))
            .border(1.dp, oak.border, RoundedCornerShape(OakRadius.md))
            .background(MaterialTheme.colorScheme.surface)
            .horizontalScroll(rememberScrollState()),
    ) {
        TableGrid(
            cols = cols,
            cellCap = 240.dp,
        ) {
            allRows.forEachIndexed { rowIndex, row ->
                val isHeader = rowIndex == 0
                val background = when {
                    isHeader -> oak.textStrong.copy(alpha = 0.06f)
                    rowIndex % 2 == 0 -> Color.Transparent
                    else -> oak.textStrong.copy(alpha = 0.04f)
                }
                for (c in 0 until cols) {
                    val value = row.getOrNull(c) ?: ""
                    val align = table.alignments.getOrNull(c) ?: MdColumnAlignment.Leading
                    Box(
                        modifier = Modifier
                            .background(background)
                            .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                        contentAlignment = when (align) {
                            MdColumnAlignment.Leading -> Alignment.CenterStart
                            MdColumnAlignment.Center -> Alignment.Center
                            MdColumnAlignment.Trailing -> Alignment.CenterEnd
                        },
                    ) {
                        Text(
                            text = parseInline(value, linkColor = oak.azure),
                            style = if (isHeader) {
                                MaterialTheme.typography.labelMedium.copy(fontWeight = FontWeight.SemiBold)
                            } else {
                                MaterialTheme.typography.bodyMedium
                            },
                            color = if (isHeader) oak.textStrong else oak.text,
                            textAlign = when (align) {
                                MdColumnAlignment.Leading -> TextAlign.Start
                                MdColumnAlignment.Center -> TextAlign.Center
                                MdColumnAlignment.Trailing -> TextAlign.End
                            },
                        )
                    }
                }
            }
        }
    }
}

/**
 * A fixed grid: [content] must emit exactly `rows * [cols]` children in row-major
 * order. Column widths are the max content width in that column (each cell capped at
 * [cellCap] so one long cell can't blow out the layout); every cell then fills its
 * column width and its row height, so backgrounds tile the whole grid and rows align.
 */
@Composable
private fun TableGrid(
    cols: Int,
    cellCap: Dp,
    content: @Composable () -> Unit,
) {
    Layout(content = content) { measurables, constraints ->
        if (measurables.isEmpty()) return@Layout layout(0, 0) {}
        val capPx = cellCap.roundToPx()
        val rowCount = (measurables.size + cols - 1) / cols

        // Pass 1 — natural width (capped) → per-column max width.
        val colWidths = IntArray(cols)
        measurables.forEachIndexed { idx, m ->
            val p = m.measure(Constraints(maxWidth = capPx))
            val c = idx % cols
            if (p.width > colWidths[c]) colWidths[c] = p.width
        }

        // Pass 2 — fixed column width → per-row max height.
        val rowHeights = IntArray(rowCount)
        measurables.forEachIndexed { idx, m ->
            val c = idx % cols
            val w = colWidths[c]
            val p = m.measure(Constraints(minWidth = w, maxWidth = w))
            val r = idx / cols
            if (p.height > rowHeights[r]) rowHeights[r] = p.height
        }

        // Pass 3 — fixed width + fixed row height so each cell fills its grid slot.
        val placeables = measurables.mapIndexed { idx, m ->
            val c = idx % cols
            val r = idx / cols
            val w = colWidths[c]
            val h = rowHeights[r]
            m.measure(Constraints(minWidth = w, maxWidth = w, minHeight = h, maxHeight = h))
        }

        val totalWidth = colWidths.sum()
        val totalHeight = rowHeights.sum()
        layout(totalWidth, totalHeight) {
            var y = 0
            for (r in 0 until rowCount) {
                var x = 0
                for (c in 0 until cols) {
                    val idx = r * cols + c
                    if (idx < placeables.size) {
                        placeables[idx].placeRelative(x, y)
                    }
                    x += colWidths[c]
                }
                y += rowHeights[r]
            }
        }
    }
}

File: ios/OakApp/UI/MarkdownBlockView.swift
import SwiftUI

/// Renders block-level Markdown — headings, lists, GFM tables, fenced code,
/// blockquotes, thematic breaks — natively, using ``MarkdownBlocks/parse(_:)`` to
/// split the source and the inline ``MarkdownText`` primitive to render the
/// bold/italic/`code`/link markup *inside* each block. This is the block-aware
/// counterpart to ``MarkdownText`` (which stays inline-only); use it wherever an
/// `OakAnswer` field can carry block structure — `answer_markdown`,
/// `reasoning_markdown`, and the live streaming bubble.
///
/// **Streaming-safe by construction:** the splitter degrades a half-formed trailing
/// block (an open code fence, a header without its separator) to a plain
/// inline-parsed paragraph, so re-rendering on every delta never drops content and
/// never flickers structure it can't yet resolve.
///
/// Typography/color inherit from the environment for prose (paragraphs, list items,
/// table cells) so a caller's `.font(…)`/`.foregroundStyle(…)` and Dynamic Type
/// flow through; only headings, code, and table chrome set their own type/wash.
struct MarkdownBlockView: View {
  private let markdown: String
  private let precomputedBlocks: [MarkdownBlock]?

  /// Text-like, unlabeled initializer mirroring ``MarkdownText``.
  init(_ markdown: String) {
    self.markdown = markdown
    self.precomputedBlocks = nil
  }

  /// Renders a pre-parsed block slice. Used by ``AnswerCardView`` to render
  /// the blocks after the lead paragraph without re-parsing the full source.
  init(blocks: [MarkdownBlock]) {
    self.markdown = ""
    self.precomputedBlocks = blocks
  }

  var body: some View {
    let blocks = precomputedBlocks ?? MarkdownBlocks.parse(markdown)
    VStack(alignment: .leading, spacing: 8) {
      ForEach(Array(blocks.enumerated()), id: \.offset) { _, block in
        view(for: block)
          .frame(maxWidth: .infinity, alignment: .leading)
      }
    }
  }

  @ViewBuilder
  private func view(for block: MarkdownBlock) -> some View {
    switch block {
    case let .heading(level, text):
      MarkdownText(text)
        .font(headingFont(level))
        .fixedSize(horizontal: false, vertical: true)
    case let .paragraph(text):
      MarkdownText(text)
        .fixedSize(horizontal: false, vertical: true)
    case let .list(items):
      MarkdownListView(items: items)
    case let .table(table):
      MarkdownTableView(table: table)
    case let .codeBlock(_, code):
      MarkdownCodeBlockView(code: code)
    case let .blockquote(text):
      MarkdownBlockquoteView(text: text)
    case .thematicBreak:
      Divider().overlay(Theme.separator)
    }
  }

  /// Heading type ramp, scaled to chat-card sizing (h1 is a section title, not a
  /// hero). `display(_:)` is rounded + semibold — Oak's heading voice. Color is
  /// left to inherit so a heading inside muted reasoning text stays muted.
  private func headingFont(_ level: Int) -> Font {
    switch level {
    case 1: return Theme.display(.title3)
    case 2: return Theme.display(.headline)
    case 3: return Theme.display(.subheadline)
    default: return Theme.display(.footnote)
    }
  }
}

// MARK: - Lists

/// An ordered/unordered list with a single supported nesting level. Bullets use a
/// mid-dot; ordered items keep their parsed ordinal. Nested (level-1) items indent.
private struct MarkdownListView: View {
  let items: [MarkdownListItem]

  var body: some View {
    VStack(alignment: .leading, spacing: 4) {
      ForEach(Array(items.enumerated()), id: \.offset) { _, item in
        HStack(alignment: .firstTextBaseline, spacing: 8) {
          Text(marker(item))
            .monospacedDigit()
            .foregroundStyle(Theme.textSecondary)
          MarkdownText(item.text)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.leading, CGFloat(item.level) * 18)
      }
    }
  }

  private func marker(_ item: MarkdownListItem) -> String {
    item.ordered ? "\(item.number)." : (item.level == 0 ? "•" : "◦")
  }
}

// MARK: - Table

/// A GFM table rendered as a native, horizontally-scrollable grid, matching the
/// `CandidatesTableView` conventions (header wash, full-width hairline, zebra rows,
/// trailing scroll fade). Cells are inline-parsed so bold/`code`/links inside a
/// cell still render.
private struct MarkdownTableView: View {
  let table: MarkdownTable

  var body: some View {
    ScrollView(.horizontal, showsIndicators: true) {
      Grid(alignment: .leading, horizontalSpacing: 0, verticalSpacing: 0) {
        GridRow {
          ForEach(Array(table.header.enumerated()), id: \.offset) { index, heading in
            cell(background: headerBackground, alignment: alignment(index)) {
              MarkdownText(heading)
                .font(Theme.body(.caption).weight(.semibold))
                .fixedSize(horizontal: false, vertical: true)
            }
          }
        }
        GridRow {
          Rectangle()
            .fill(Theme.separator)
            .frame(height: 1)
            .gridCellUnsizedAxes(.horizontal)
            .gridCellColumns(max(table.header.count, 1))
        }
        ForEach(Array(table.rows.enumerated()), id: \.offset) { rowIndex, row in
          GridRow {
            ForEach(Array(row.enumerated()), id: \.offset) { colIndex, value in
              cell(background: rowBackground(rowIndex), alignment: alignment(colIndex)) {
                MarkdownText(value)
                  .font(Theme.body(.subheadline))
                  .fixedSize(horizontal: false, vertical: true)
              }
            }
          }
        }
      }
    }
    .background(Theme.surface)
    .oakCard(radius: Theme.Radius.md)
    .overlay(alignment: .trailing) { scrollFade }
  }

  private func alignment(_ column: Int) -> Alignment {
    guard column < table.alignments.count else { return .leading }
    switch table.alignments[column] {
    case .leading: return .leading
    case .center: return .center
    case .trailing: return .trailing
    }
  }

  private var scrollFade: some View {
    LinearGradient(
      colors: [Theme.surface.opacity(0), Theme.surface],
      startPoint: .leading,
      endPoint: .trailing
    )
    .frame(width: 24)
    .allowsHitTesting(false)
  }

  private var headerBackground: Color { Theme.textPrimary.opacity(0.06) }

  private func rowBackground(_ index: Int) -> Color {
    index.isMultiple(of: 2) ? Color.clear : Theme.textPrimary.opacity(0.04)
  }

  @ViewBuilder
  private func cell<Content: View>(
    background: Color,
    alignment: Alignment,
    @ViewBuilder content: () -> Content
  ) -> some View {
    content()
      .padding(.horizontal, 12)
      .padding(.vertical, 8)
      .frame(maxWidth: .infinity, alignment: alignment)
      .background(background)
  }
}

// MARK: - Code block

/// A fenced code block: monospaced, on a raised surface, horizontally scrollable so
/// long lines never force the page to scroll sideways. No syntax highlighting.
private struct MarkdownCodeBlockView: View {
  let code: String

  var body: some View {
    ScrollView(.horizontal, showsIndicators: false) {
      Text(code)
        .font(Theme.mono(.footnote))
        .foregroundStyle(Theme.textPrimary)
        .textSelection(.enabled)
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
    .background(Theme.surface)
    .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous))
    .overlay(
      RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
        .strokeBorder(Theme.separator, lineWidth: 1)
    )
  }
}

// MARK: - Blockquote

/// A blockquote: an accent rule on the leading edge with muted, inline-parsed body
/// text.
private struct MarkdownBlockquoteView: View {
  let text: String

  var body: some View {
    HStack(alignment: .top, spacing: 10) {
      RoundedRectangle(cornerRadius: 2, style: .continuous)
        .fill(Theme.accent.opacity(0.5))
        .frame(width: 3)
      MarkdownText(text)
        .foregroundStyle(Theme.textSecondary)
        .fixedSize(horizontal: false, vertical: true)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
    .fixedSize(horizontal: false, vertical: true)
  }
}

#if DEBUG
#Preview("MarkdownBlockView — mixed document") {
  ScrollView {
    MarkdownBlockView(
      """
      ## Best moveset for Garchomp

      Garchomp wants a **physical** set. Key moves:

      - Dragon Claw — reliable STAB
      - Earthquake — coverage
        - hits Steel/Rock hard
      1. Lead with it
      2. Pivot out

      | Move | Type | BP |
      | --- | --- | ---: |
      | Dragon Claw | Dragon | 80 |
      | Earthquake | Ground | 100 |

      > Speed tiers matter — `108` base outpaces most walls.

      ```
      252 Atk Garchomp Earthquake
      ```

      ---

      See [Bulbapedia](https://bulbapedia.bulbagarden.net) for details.
      """
    )
    .font(Theme.body(.body))
    .foregroundStyle(Theme.textPrimary)
    .padding()
  }
}
#endif
